# Машинное обучение. Лабораторная работа 1

Выполнил: _Коростелев Дмитрий Васильевич_

Группа:   _М8О-308Б-18_

## Настройка среды

In [1]:
%pip install numpy
%pip install pandas
%pip install sklearn

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


## Задача и датасет

Датасет взят с сайта : http://archive.ics.uci.edu/ml/datasets/Mushroom

В базе данных представлена информацию о видах грибов, их свойствах и особенностях. По описанию гриба при помощи искусственного интеллекта будем решать задачу классификации грибов на ядовитые и съедобные

## Датасет

In [2]:
import pandas as pds

df = pds.read_csv('agaricus-lepiota.data')

In [3]:
df

edible cap-shape cap-surface cap-color bruises odor gill-attachment  \
0         p         x           s         n       t    p               f   
1         e         x           s         y       t    a               f   
2         e         b           s         w       t    l               f   
3         p         x           y         w       t    p               f   
4         e         x           s         g       f    n               f   
5         e         x           y         y       t    a               f   
6         e         b           s         w       t    a               f   
7         e         b           y         w       t    l               f   
8         p         x           y         w       t    p               f   
9         e         b           s         y       t    a               f   
10        e         x           y         y       t    l               f   
11        e         x           y         y       t    a               f   
12        e         b           s         y       t    a               f   
13        p         x           y         w       t    p               f   
14        e         x           f         n       f    n               f   
15        e         s           f         g       f    n               f   
16        e         f           f         w       f    n               f   
17        p         x           s         n       t    p               f   
18        p         x           y         w       t    p               f   
19        p         x           s         n       t    p               f   
20        e         b           s         y       t    a               f   
21        p         x           y         n       t    p               f   
22        e         b           y         y       t    l               f   
23        e         b           y         w       t    a               f   
24        e         b           s         w       t    l               f   
25        p         f           s         w       t    p               f   
26        e         x           y         y       t    a               f   
27        e         x           y         w       t    l               f   
28        e         f           f         n       f    n               f   
29        e         x           s         y       t    a               f   
...     ...       ...         ...       ...     ...  ...             ...   
8094      e         b           s         g       f    n               f   
8095      p         x           y         c       f    m               f   
8096      e         k           f         w       f    n               f   
8097      p         k           y         n       f    s               f   
8098      p         k           s         e       f    y               f   
8099      e         k           f         w       f    n               f   
8100      e         f           s         n       f    n               a   
8101      p         k           s         e       f    s               f   
8102      e         x           s         n       f    n               a   
8103      e         k           s         n       f    n               a   
8104      e         k           s         n       f    n               a   
8105      e         k           s         n       f    n               a   
8106      e         k           s         n       f    n               a   
8107      e         x           s         n       f    n               a   
8108      p         k           y         e       f    y               f   
8109      e         b           s         w       f    n               f   
8110      e         x           s         n       f    n               a   
8111      e         k           s         w       f    n               f   
8112      e         k           s         n       f    n               a   
8113      p         k           y         e       f    y               f   
8114      p         f           y         c       f  

## Описание атрибутов и возможных значений: 

0. edible : p=poisonous, e=edible
1. cap-shape: bell=b,conical=c,convex=x,flat=f, knobbed=k,sunken=s
2. cap-surface: fibrous=f,grooves=g,scaly=y,smooth=s
3. cap-color: brown=n,buff=b,cinnamon=c,gray=g,green=r, pink=p,purple=u,red=e,white=w,yellow=y
4. bruises?: bruises=t,no=f
5. odor: almond=a,anise=l,creosote=c,fishy=y,foul=f, musty=m,none=n,pungent=p,spicy=s
6. gill-attachment: attached=a,descending=d,free=f,notched=n
7. gill-spacing: close=c,crowded=w,distant=d
8. gill-size: broad=b,narrow=n
9. gill-color: black=k,brown=n,buff=b,chocolate=h,gray=g, green=r,orange=o,pink=p,purple=u,red=e, white=w,yellow=y
10. stalk-shape: enlarging=e,tapering=t
11. stalk-root: bulbous=b,club=c,cup=u,equal=e, rhizomorphs=z,rooted=r,missing=?
12. stalk-surface-above-ring: fibrous=f,scaly=y,silky=k,smooth=s
13. stalk-surface-below-ring: fibrous=f,scaly=y,silky=k,smooth=s
14. stalk-color-above-ring: brown=n,buff=b,cinnamon=c,gray=g,orange=o, pink=p,red=e,white=w,yellow=y
15. stalk-color-below-ring: brown=n,buff=b,cinnamon=c,gray=g,orange=o, pink=p,red=e,white=w,yellow=y
16. veil-type: partial=p,universal=u
17. veil-color: brown=n,orange=o,white=w,yellow=y
18. ring-number: none=n,one=o,two=t
19. ring-type: cobwebby=c,evanescent=e,flaring=f,large=l, none=n,pendant=p,sheathing=s,zone=z
20. spore-print-color: black=k,brown=n,buff=b,chocolate=h,green=r, orange=o,purple=u,white=w,yellow=y
21. population: abundant=a,clustered=c,numerous=n, scattered=s,several=v,solitary=y
22. habitat: grasses=g,leaves=l,meadows=m,paths=p, urban=u,waste=w,woods=d

Разделим датасет на обучающую и тестовую выборки. Так как записей в датасете - 8123, разделим датасет в соотношении 3:1 

In [4]:
separate_index = int(len(df)/4.0)
print(separate_index)

2031


In [5]:
df_test = df.iloc[0:separate_index]
df_learn = df.iloc[separate_index:]

# Алгоритм KNN

## Подготовка датасета

Реализуем алгоритм KNN, сначала сопоставим всем качественным признакам числовые значения и будем работать с числовыми значениями.

In [6]:
edible_dict = { 'p' : 0, 'e' : 1 }
cap_shape_dict = { 'b' : 0, 'c' : 1, 'x' : 2, 'f' : 3, 'k' : 4, 's' : 5 }
cap_surface_dict = { 'f' : 0, 'g' : 1, 'y' : 2, 's' : 3 }
cap_color_dict = { 'n' : 0, 'b' : 1, 'c' : 2, 'g' : 3, 'r' : 4, 'p' : 5, 'u' : 6, 'e' : 7, 'w' : 8, 'y' : 9 }
bruises_dict = { 't' : 0, 'f' : 1 }
odor_dict = { 'a' : 0, 'l' : 1, 'c' : 2, 'y' : 3, 'f' : 4, 'm' : 5, 'n' : 6, 'p' : 7, 's' : 8 }
gill_attachment_dict = { 'a' : 0, 'd' : 1, 'f' : 2, 'n' : 3 }
gill_spacing_dict = { 'c' : 0, 'w' : 1, 'd' : 2 }
gill_size_dict = { 'b' : 0, 'n' : 1 }
gill_color_dict = { 'k' : 0, 'n' : 1, 'b' : 2, 'h' : 3, 'g' : 4, 'r' : 5, 'o' : 6, 'p' : 7, 'u' : 8, 'e' : 9, 'w' : 10, 'y' : 11 }
stalk_shape_dict = { 'e' : 0, 't' : 1 }
stalk_root_dict = { 'b' : 0, 'c' : 1, 'u' : 2, 'e' : 3, 'z' : 4, 'r' : 5, '?' : 6 }
stalk_surface_above_ring_dict = { 'f' : 0, 'y' : 1, 'k' : 2, 's' : 3 }
stalk_surface_below_ring_dict = { 'f' : 0, 'y' : 1, 'k' : 2, 's' : 3 }
stalk_color_above_ring_dict = { 'n' : 0, 'b' : 1, 'c' : 2, 'g' : 3, 'o' : 4, 'p' : 5, 'e' : 6, 'w' : 7, 'y' : 8 }
stalk_color_below_ring_dict = { 'n' : 0, 'b' : 1, 'c' : 2, 'g' : 3, 'o' : 4, 'p' : 5, 'e' : 6, 'w' : 7, 'y' : 8 }
veil_type_dict = { 'p' : 0, 'u' : 1 }
veil_color_dict = { 'n' : 0, 'o' : 1, 'w' : 2, 'y' : 3 }
ring_number_dict = { 'n' : 0, 'o' : 1, 't' : 2 }
ring_type_dict = { 'c' : 0, 'e' : 1, 'f' : 2, 'l' : 3, 'n' : 4, 'p' : 5, 's' : 6, 'z' : 7 }
spore_print_color_dict = { 'k' : 0, 'n' : 1, 'b' : 2, 'h' : 3, 'r' : 4, 'o' : 5, 'u' : 6, 'w' : 7, 'y' : 8 }
population_dict = { 'a' : 0, 'c' : 1, 'n' : 2, 's' : 3, 'v' : 4, 'y' : 5 }
habitat_dict = { 'g' : 0, 'l' : 1, 'm' : 2, 'p' : 3, 'u' : 4, 'w' : 5, 'd' : 6 }

def replace_attribute_on_number(row):
    row['edible'] = edible_dict[row['edible']] # 1
    row['cap-shape'] = cap_shape_dict[row['cap-shape']] # 2
    row['cap-surface'] = cap_surface_dict[row['cap-surface']] # 3
    row['cap-color'] = cap_color_dict[row['cap-color']] # 4
    row['bruises'] = bruises_dict[row['bruises']] # 5
    row['odor'] = odor_dict[row['odor']] # 6
    row['gill-attachment'] = gill_attachment_dict[row['gill-attachment']] # 7
    row['gill-spacing'] = gill_spacing_dict[row['gill-spacing']] # 8
    row['gill-size'] = gill_size_dict[row['gill-size']] # 9
    row['gill-color'] = gill_color_dict[row['gill-color']] # 10
    row['stalk-shape'] = stalk_shape_dict[row['stalk-shape']] # 11
    row['stalk-root'] = stalk_root_dict[row['stalk-root']] # 12
    row['stalk-surface-above-ring'] = stalk_surface_above_ring_dict[row['stalk-surface-above-ring']] # 13
    row['stalk-surface-below-ring'] = stalk_surface_below_ring_dict[row['stalk-surface-below-ring']] # 14
    row['stalk-color-above-ring'] = stalk_color_above_ring_dict[row['stalk-color-above-ring']] # 15
    row['stalk-color-below-ring'] = stalk_color_below_ring_dict[row['stalk-color-below-ring']] # 16
    row['veil-type'] = veil_type_dict[row['veil-type']] # 17
    row['veil-color'] = veil_color_dict[row['veil-color']] # 18
    row['ring-number'] = ring_number_dict[row['ring-number']] # 19
    row['ring-type'] = ring_type_dict[row['ring-type']] # 20
    row['spore-print-color'] = spore_print_color_dict[row['spore-print-color']] # 21
    row['population'] = population_dict[row['population']] # 22
    row['habitat'] = habitat_dict[row['habitat']] # 23


def change_atr_to_numbers_in_df(df):
    df_copy = df.copy()
    for index, row in df_copy.iterrows():
        replace_attribute_on_number(row)
    return df_copy

df_learn_numeric = change_atr_to_numbers_in_df(df_learn)


In [7]:
df_learn_numeric

edible cap-shape cap-surface cap-color bruises odor gill-attachment  \
2031      1         3           3         0       1    6               2   
2032      1         2           0         7       0    6               2   
2033      1         3           2         0       0    6               2   
2034      1         2           2         3       0    6               2   
2035      1         2           3         3       1    6               2   
2036      1         2           3         0       1    6               2   
2037      1         2           0         3       1    6               2   
2038      1         2           3         3       1    6               2   
2039      1         2           2         7       0    6               2   
2040      1         3           0         3       0    6               2   
2041      1         3           0         0       0    6               2   
2042      1         2           0         0       1    6               2   
2043      1         3           3         0       1    6               2   
2044      1         2           0         8       1    6               2   
2045      1         2           2         9       0    0               2   
2046      1         2           0         7       0    6               2   
2047      1         2           0         0       1    6               2   
2048      1         2           2         0       0    6               2   
2049      1         3           0         3       0    6               2   
2050      1         2           0         8       1    6               2   
2051      1         2           2         8       0    0               2   
2052      1         2           3         0       1    6               2   
2053      1         3           3         8       1    6               2   
2054      1         3           0         0       0    6               2   
2055      1         3           0         8       1    6               2   
2056      1         2           0         3       0    6               2   
2057      1         2           0         0       0    6               2   
2058      1         2           0         8       1    6               2   
2059      1         2           3         8       0    0               2   
2060      1         3           3         3       1    6               2   
...     ...       ...         ...       ...     ...  ...             ...   
8094      1         0           3         3       1    6               2   
8095      0         2           2         2       1    5               2   
8096      1         4           0         8       1    6               2   
8097      0         4           2         0       1    8               2   
8098      0         4           3         7       1    3               2   
8099      1         4           0         8       1    6               2   
8100      1         3           3         0       1    6               0   
8101      0         4           3         7       1    8               2   
8102      1         2           3         0       1    6               0   
8103      1         4           3         0       1    6               0   
8104      1         4           3         0       1    6               0   
8105      1         4           3         0       1    6               0   
8106      1         4           3         0       1    6               0   
8107      1         2           3         0       1    6               0   
8108      0         4           2         7       1    3               2   
8109      1         0           3         8       1    6               2   
8110      1         2           3         0       1    6               0   
8111      1         4           3         8       1    6               2   
8112      1         4           3         0       1    6               0   
8113      0         4           2         7       1    3               2   
8114      0         3           2         2       1  

Получим ядовитые и съедобные грибы

In [8]:
df_learn_poisonous = df_learn_numeric[df_learn_numeric['edible'] == 0]
df_learn_edible =    df_learn_numeric[df_learn_numeric['edible'] == 1]

Проверим, что разделение прошло успешно

In [9]:
print((len(df_learn_edible) + len(df_learn_poisonous) == len(df_learn_numeric)))

True


## Функционал

Датасет подготовлен для обучения модели. Введем несколько дополнительную функцию - эвклидово расстояние по которому будем рассчитывать расстояние

In [10]:
from math import sqrt
def euclidean(v1, v2):
    return sqrt(sum([(x - y)**2 for x, y in zip(v1, v2)]))

## Обучение модели

Обучать модель не нужно, knn достаточно предоставить базу данных с заранее известными ответами. Ответ будем получать сравнивая кол-во соседей заданной точки с заранее известными и затем выводить ответ в зависимости от того каких соседей больше.

### Тестирование

In [32]:
def is_edible_knn(row, k = 8):
    edible_count = 0
    poison_count = 0
    result_list = list()
    while len(result_list) != k:
        min_value = 1000000
        min_row = None
        for index, row_iter in df_learn_numeric.iterrows():
            if euclidean(row[1:], row_iter[1:]) < min_value and not index in result_list:
                min_row = row_iter
                result_list.append(index)
                break
    for index in result_list:
        if df_learn_numeric.loc[index]['edible'] == 1:
            edible_count += 1
        else:
            poison_count += 1
    return edible_count > poison_count
        
def test_knn(df):
    df_numeric = change_atr_to_numbers_in_df(df)
    success_count = 0
    for index, row in df_numeric.iterrows():
        real_answer = (row['edible'] == 1)
        algo_answer = is_edible_knn(row)
        if real_answer == algo_answer:
            success_count += 1
    return success_count/len(df)
    
print("Точность алгоритма: {}%".format(test_knn(df_test)*100))

Точность алгоритма: 88.18316100443131%


## Наивный байесовский классификатор

Реализуем простейший байесовский классификатор

Для начала посчитаем сколько в обучающей выборке встречаются ядовитые и съедобвные грибы

In [13]:
count_poisonous = len(df_learn_poisonous)
count_edible = len(df_learn_edible)
print("Ядовитых грибов: {}, съедобных: {}".format(count_poisonous,count_edible))

Ядовитых грибов: 3676, съедобных: 2417


Посчитаем вероятности встретить каждый признак в каждом классе

In [14]:
def get_dict_probabilities(df):
    res = list()
    df_lists = df.values.tolist()
    for i in range(1,23):
        atr_probabilities = dict()
        for llist in df_lists:
            if llist[i] in atr_probabilities:
                atr_probabilities[llist[i]] += 1
            else :
                atr_probabilities[llist[i]] = 1
        for item in atr_probabilities:
            atr_probabilities[item] /= len(df)
        res.append(atr_probabilities)
    return res

dict_poisonous = get_dict_probabilities(df_learn_poisonous)
dict_edible = get_dict_probabilities(df_learn_edible)

### Тестирование

In [15]:
def is_edible_bayes(row):
    list_row = row.tolist()
    sum_poisonous = 0
    sum_edible = 0
    for i in range(1,23):
        if list_row[i] in dict_poisonous[i-1]:
            sum_poisonous += dict_poisonous[i-1][list_row[i]]
        if list_row[i] in dict_edible[i-1]:
            sum_edible += dict_edible[i-1][list_row[i]]
    return sum_edible > sum_poisonous

def test_bayes(df):
    df_numeric = change_atr_to_numbers_in_df(df)
    success_count = 0
    for index, row in df_numeric.iterrows():
        real_answer = (row['edible'] == 1)
        algo_answer = is_edible_bayes(row)
        if real_answer == algo_answer:
            success_count += 1
    return success_count/len(df)

print("Точность алгоритма: {}%".format(test_bayes(df_test)*100))

Точность алгоритма: 88.23239783357951%


## Сравнение с реализацией sklearn

### Алгоритм KNN

Подготовим данные для алгоритма

In [16]:
df_sklearn_train_data = list()
df_sklearn_train_ans = list()
for index, row in df_learn_poisonous.iterrows():
    df_sklearn_train_data.append(row.values[1:])
    df_sklearn_train_ans.append(row.values[0])
for index, row in df_learn_edible.iterrows():
    df_sklearn_train_data.append(row.values[1:])
    df_sklearn_train_ans.append(row.values[0])
    
df_sklearn_test_data = list()
df_sklearn_test_ans = list()
for index, row in change_atr_to_numbers_in_df(df_test).iterrows():
    df_sklearn_test_data.append(row.values[1:])
    df_sklearn_test_ans.append(row.values[0])

Проводим тестирование

In [33]:
from sklearn.neighbors import KNeighborsClassifier

nbrs = KNeighborsClassifier(n_neighbors=7, algorithm='auto').fit(df_sklearn_train_data, df_sklearn_train_ans)

print("Точность алгоритма : {}%".format(nbrs.score(df_sklearn_test_data, df_sklearn_test_ans)*100))

Точность алгоритма : 75.23387493845397%


### Байесовский классификатор

Проводим тестирование

In [34]:
from sklearn.naive_bayes import GaussianNB

gnb = GaussianNB().fit(df_sklearn_train_data, df_sklearn_train_ans)

print("Точность алгоритма : {}%".format(gnb.score(df_sklearn_test_data, df_sklearn_test_ans)*100))

Точность алгоритма : 49.67996061053668%


## Выводы

В ходе лабораторной работы были реализованы простейшие классификаторы KNN и байесовкий с приблизительной точностью для выбранного мной датасета в 88% каждый. Сравнил собственную реализацию с реализациями в библиотеке sklearn. Точность KNN собственной реализации оказалась чуть выше чем в sklearn, алгоритм байесовского классификатора в sklearn показал довольно низкие результаты, скорее всего это связано с тем, что данный 

В ходе выполнения работы смог убедится на практике в том, что при помощи математики, вычислительной техники и данных можно находить различные корреляции между метриками, которые не поддаются обычному разъяснению для единичных сущностей, иначе говоря, на основе статистики можно с большой долей вероятности предсказать некоторые свойства выбранного объекта. 